### PR TITLE
[WIP] migrate to classmethod get_build_plan()

### PR DESCRIPTION
**Pull request by tigarmo**
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: rockcraft/models/project.py [hunk 1]
```diff
@@ -431,12 +431,17 @@ def generate_metadata(
...
+    def get_build_plan(cls, data: Dict[str, Any]) -> List[BuildInfo]:
         """Obtain the list of architectures and bases from the project file."""
         build_infos: List[BuildInfo] = []
-        base = self.effective_base
 
-        for platform_entry, platform in self.platforms.items():
+        platforms = cls._validate_all_platforms(data["platforms"])
+        effective_base = data["build-base"] if "build-base" in data else data["base"]
+        name, channel = effective_base.replace(":", "@").split("@")
+        base = bases.BaseName(name, channel)
```
Comment (cmatsuoka) [inline rockcraft/models/project.py:442]: Maybe craft-application could provide a small model containing just platforms, base and build-base with this logic built-in?

File: docs/how-to/code/install-slice/rockcraft.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 name: custom-openssl-rock
 base: bare
-build_base: "ubuntu@22.04"
```
Comment (tigarmo) [inline docs/how-to/code/install-slice/rockcraft.yaml:3]: a side-effect of this change is that I found out that we apparently accept `build_base` as well as `build-base`
weird

File: rockcraft/services/image.py [hunk 3]
```diff
@@ -48,8 +48,9 @@ def __init__(
         project: models.Project,
         work_dir: Path,
         build_for: str,
+        **kwargs,
     ):
-        super().__init__(app, services, project=project)
+        super().__init__(app, services, project=project, **kwargs)
```
Comment (tigarmo) [inline rockcraft/services/image.py:53]: I thought it was very "bureaucratic" to have to add the new parameter to all services, even those that don't use it (like this ImageService)

File: tests/unit/test_project.py [hunk 4]
```diff
@@ -625,5 +625,5 @@ def test_project_yaml(yaml_loaded_data):
 )
 def test_project_get_build_plan(yaml_loaded_data, platforms, expected_build_infos):
     yaml_loaded_data["platforms"] = platforms
-    project = Project.unmarshal(yaml_loaded_data)
-    assert project.get_build_plan() == expected_build_infos
+    build_plan = Project.get_build_plan(yaml_loaded_data)
```
Comment (tigarmo) [inline tests/unit/test_project.py:628]: a loss of usability here is that I can no longer call `project.get_build_plan()` without parameters even if I already have a project (with its base and platforms already validated and set)

**Pull request closed without merging**